src/solidity/utils.ts

```diff
@@ -24,52 +24,65 @@ export function findImports(data: ImportFile): string[] {
   workingDir: string,
-  resolver: ResolverEngine<ImportFile> = SolidityImportResolver(),
-): Promise<ImportFile[]> {
-  let result: ImportFile[] = [];
-  let queue: { cwd: string; file: string }[] = [];
+  resolver: ResolverEngine<ImportFile>,
+): Promise<ImportTreeNode[]> {
+  let result: ImportTreeNode[] = [];
   let alreadyImported = new Set();
 
-  // solc resolves relative paths starting from current file's path, so if we leave relative path then
-  // imported path "../../a/b/c.sol" from file "file.sol" resolves to a/b/c.sol, which is wrong.
-  // we start from file;s absolute path so relative path can resolve correctly
-  const absoluteWhats = whats.map(what => path.resolve(workingDir, what));
-  for (const absWhat of absoluteWhats) {
-    queue.push({ cwd: workingDir, file: absWhat });
-    alreadyImported.add(solidifyName(absWhat));
-  }
-  while (queue.length > 0) {
-    const fileData = queue.shift()!;
-    const resolvedFile: ImportFile = await resolver.require(fileData.file, fileData.cwd);
-    const foundImports = findImports(resolvedFile);
-
-    // if imported path starts with '.' we assume it's relative and return it's absolute path
-    // if not - return the same name it was imported with
-    if (fileData.file[0] === ".") {
-      result.push(resolvedFile);
-    } else {
-      result.push({ url: fileData.file, source: resolvedFile.source });
+  async function dfs(file: { searchCwd: string; uri: string }): Promise<string> {
+    const url = await resolver.resolve(file.uri, file.searchCwd);
+    if (alreadyImported.has(url)) {
+      return url;
     }
 
-    const fileParentDir = path.dirname(resolvedFile.url);
-    for (const i in foundImports) {
-      const solidifiedName: string = solidifyName(foundImports[i]);
-      if (!alreadyImported.has(solidifiedName)) {
-        alreadyImported.add(solidifiedName);
-        queue.push({ cwd: fileParentDir, file: foundImports[i] });
-      }
+    const resolvedFile = await resolver.require(file.uri, file.searchCwd);
+
+    alreadyImported.add(url);
+
+    const foundImportURIs = findImports(resolvedFile);
+
+    const fileNode: ImportTreeNode = { uri: file.uri, imports: [], ...resolvedFile };
+
+    const resolvedCwd = path.dirname(url);
+    for (const importUri of foundImportURIs) {
+      const importUrl = await dfs({ searchCwd: resolvedCwd, uri: importUri });
+      fileNode.imports.push({ uri: importUri, url: importUrl });
     }
+
+    result.push(fileNode);
+    return resolvedFile.url;
   }
 
+  await Promise.all(whats.map(what => dfs({ searchCwd: workingDir, uri: what })));
+
   return result;
 }
+
+export async function gatherSources(
+  whats: string[],
+  workingDir: string,
+  resolver: ResolverEngine<ImportFile> = SolidityImportResolver(),
+): Promise<ImportFile[]> {
+  return (await traverseSources(whats, workingDir, resolver)).map(node => {
+    return { url: node.url, source: node.source };
+  });
+}
+
+export async function gatherSourcesAndCanonizeImports(
```

Comment:
Some note on the fact that this is the _magic_ function that messes with contents of the sources?